### PR TITLE
[CBRD-22955] Filtered apply after master crash

### PR DESCRIPTION
**Pull request by radudoros**
http://jira.cubrid.org/browse/CBRD-22955

Comprising changes:
- Find all mvccids of active transactions
- Fetch stream from master starting with m_active_stream_position
- Apply fetched stream and skip stream_entries having mvccids not in the gathered mvccids of active transactions

### CODE REVIEW

File: src/replication/log_generator.hpp [hunk 1]
```diff
@@ -84,6 +84,8 @@ namespace cubreplication
 
       bool m_has_stream;
       bool m_is_row_replication_disabled;
+      cubstream::stream_position gc_end_position;
```
Comment (ribram) [inline src/replication/log_generator.hpp]: Add `m_` prefix. 

Is this temporary field until group complete is merged?

File: src/replication/replication_node_manager.cpp [hunk 2]
```diff
@@ -45,12 +46,24 @@ namespace cubreplication
     {
       g_hostname = server_name;
 
+      g_start_position = 0;
```
Comment (ribram) [inline src/replication/replication_node_manager.cpp]: g_start_position looks like a local variable.

File: src/replication/replication_node_manager.cpp [hunk 3]
```diff
@@ -45,12 +46,24 @@ namespace cubreplication
...
+	  g_start_position = log_Gl.hdr.m_ack_stream_position;
+	}
+
       INT64 buffer_size = prm_get_bigint_value (PRM_ID_REPL_BUFFER_SIZE);
       int num_max_appenders = log_Gl.trantable.num_total_indices + 1;
       g_stream = new cubstream::multi_thread_stream (buffer_size, num_max_appenders);
       g_stream->set_name ("repl" + g_hostname);
       g_stream->set_trigger_min_to_read_size (stream_entry::compute_header_size ());
-      g_stream->init (0);
+      g_stream->init (g_start_position);
```
Comment (ribram) [inline src/replication/replication_node_manager.cpp]: Is it correct to initialize stream with active_start_position?

File: src/replication/replication_slave_node.cpp [hunk 4]
```diff
@@ -97,30 +97,42 @@ namespace cubreplication
 	return error;
       }
 
+    cubstream::stream_position start_position = 0;
```
Comment (ribram) [inline src/replication/replication_slave_node.cpp]: setting start position code is duplicated - here and in replication_node_manager.cpp. can we avoid duplicating it?

File: src/replication/log_consumer.cpp [hunk 5]
```diff
@@ -82,6 +82,9 @@ namespace cubreplication
 
       void execute (cubthread::entry &thread_ref) final
       {
+	assert (!m_repl_stream_entries.empty () && m_repl_stream_entries.back ()->is_group_commit ());
```
Comment (ribram) [inline src/replication/log_consumer.cpp]: Are you sure the last replication object must be a group commit?

File: src/replication/log_consumer.cpp [hunk 6]
```diff
@@ -181,6 +191,42 @@ namespace cubreplication
 		break;
 	      }
 
+	    MVCCID mvccid = se->get_mvccid ();
+
+	    // apply only some selected mvccids until we reach last m_ack_stream_position, then continue normally
+	    if (is_filtered_apply_segment (se->get_stream_entry_end_position ()))
+	      {
+		_er_log_debug (ARG_FILE_LINE, "Filtered apply: Entered filtered apply segment, m_filtered_apply_end: %llu\n",
```
Comment (ribram) [inline src/replication/log_consumer.cpp]: Shouldn't group commit stream entries be skipped during filtered phase?

File: src/transaction/log_recovery.c [hunk 7]
```diff
@@ -5084,7 +5088,90 @@ log_recovery_undo (THREAD_ENTRY * thread_p)
...
+		    {
+		    case LOG_MVCC_UNDOREDO_DATA:
+		    case LOG_MVCC_DIFF_UNDOREDO_DATA:
+		      {
+			LOG_READ_ADD_ALIGN (thread_p, sizeof (LOG_RECORD_HEADER), &log_lsa, log_pgptr);
+			size_t rec_size = sizeof (LOG_REC_MVCC_UNDOREDO);
+			LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, rec_size, &log_lsa, log_pgptr);
+			LOG_REC_MVCC_UNDOREDO *mvcc_undoredo =
+			  (LOG_REC_MVCC_UNDOREDO *) ((char *) log_pgptr->area + log_lsa.offset);
+			log_Gl.m_active_mvcc_ids.insert (mvcc_undoredo->mvccid);
```
[the remark this answers is not in the document]
Comment (radudoros) [reply src/transaction/log_recovery.c]: I think m_active_tran_ids can be removed if the mvcc gathering is done at analysis. I postponed doing that. Do you think this pr is the moment we should do it?

File: src/replication/log_consumer.cpp [hunk 8]
```diff
@@ -113,6 +116,7 @@ namespace cubreplication
 	  }
 
 	(void) locator_repl_end_tran (&thread_ref, true);
+	log_Gl.hdr.m_ack_stream_position = last_applied_ack;
```
[the remark this answers is not in the document]
Comment (radudoros) [reply src/replication/log_consumer.cpp]: I think so

File: src/replication/log_consumer.cpp (hunk 6 above)
Comment (radudoros) [reply src/replication/log_consumer.cpp]: I think they do get skipped

File: src/replication/log_consumer.cpp (hunk 5 above)
Comment (radudoros) [reply src/replication/log_consumer.cpp]: Fixed :)

File: src/transaction/log_recovery.c [hunk 9]
```diff
@@ -5084,7 +5088,91 @@ log_recovery_undo (THREAD_ENTRY * thread_p)
...
 	}
 
-      assert (log_Gl.m_active_start_position != 0);
+      _er_log_debug (ARG_FILE_LINE, "HA recovery: Finished recovery of min active stream position"
+		     "m_active_start_position=%llu, m_ack_stream_position=%llu",
+		     (std::uint64_t) log_Gl.m_repl_rv.m_active_start_position,
+		     (std::uint64_t) log_Gl.hdr.m_ack_stream_position);
+    }
+
+  if (!LSA_ISNULL (&log_Gl.m_min_active_lsa))
```
Comment (eseokoh) [inline src/transaction/log_recovery.c]: Is it possible to merge this if block to the very above?

Comment (radudoros) [reply src/transaction/log_recovery.c]: removed the block above, and handled things in logtb_rv_assign_mvccid_for_undo_recovery

File: src/replication/replication_slave_node.cpp [hunk 10]
```diff
@@ -104,23 +120,22 @@ namespace cubreplication
 	return error;
       }
 
-    /* TODO[replication] : last position to be retrieved from recovery module */
-    cubstream::stream_position start_position = 0;
+    slave_control_sender *ctrl_sender = new slave_control_sender (std::move (slave_control_channel (std::move (
```
Comment (razvanradulescu) [inline src/replication/replication_slave_node.cpp]: A reader might need (as I did) a short comment to explain the slave_control_channel  constructor
(slave_control_channel  is  not a channel, it 'has' a channel) and slave_control_sender  requires a slave_control_channel.

File: src/replication/log_consumer.cpp [hunk 11]
```diff
@@ -115,6 +115,7 @@ namespace cubreplication
 	  }
 
 	(void) locator_repl_end_tran (&thread_ref, true);
+	log_Gl.hdr.m_ack_stream_position = m_gc_end_position;
```
Comment (alexstanro) [inline src/replication/log_consumer.cpp]: It may be better to update only once.

**Pull request merged**